### DISCUSSION TITLE
[General] Testing best practices

### DISCUSSION THREAD

**Discussion by GavinStein1** [zkSync-Community-Hub/zksync-developers#95] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

I am currently transferring a project from L1 to zk-sync. Previously, I would test my contracts with hardhat-ethers library and the hardhat network built into Hardhat. I would like to know if there is a similar network/functionality with zksync-era.

### What is the specific issue or error you're encountering?

The Migration guide (https://era.zksync.io/docs/tools/hardhat/migrating-to-zksync.html) was helpful, but because my project was small enough, I found it easier to create a new project with zksync-cli and paste the contracts in to the /contracts dir. My old test scripts are obviously incompatible now, and I want to know the best way to quickly test contract functionality (not on a public testnet).

### Can you share the error messages or logs you're receiving, if any?

-

### Have you made any recent changes to the contract before encountering this issue?

Contracts are working on L1 when tested on the hardhat network.

### Are there any external libraries or contracts that your contract interacts with?

-

### Can you provide the relevant portions of your contract code where the issue is occurring?

In L1, with hardhat, my test scripts looked like this to deploy and test functionality.

```
beforeEach(async () => {
    eventFactory = await ethers.deployContract("EventFactory");
    await eventFactory.waitForDeployment();
    ...
}
```

### Have you tried to isolate the problem, and if so, what were the results?

My new test script looks like this:

```
const RICH_WALLET_PK =
  '0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110';

// Deploy a contract with constructorArguments
async function deployContract(deployer: Deployer, contractName: string, constructorArguments: any[]): Promise<Contract> {
    const artifact = await deployer.loadArtifact(contractName);
    return await deployer.deploy(artifact, constructorArguments);
}
let eventFactoryContract;
beforeEach(async () => {
    const provider = Provider.getDefaultProvider();
    const wallet = new Wallet(RICH_WALLET_PK, provider);
    const deployer = new Deployer(hre, wallet);
    eventFactoryContract = await deployContract(deployer, "EventFactory", []);
});

describe("Event", function () {
    describe("Deployment", function () {
        it("Should be deployed", async function () {
            console.log(eventFactoryContract.address);
            console.log("Done");
        });
    });
});
```

### What steps have you already taken to try to resolve the issue?

The project template has the following test network if the environment variable NODE_ENV is "test":
```
{
    url: "http://localhost:3050",
    ethNetwork: "http://localhost:8545",
    zksync: true,
}
```

I am not sure how this network is activated and connected to. When I run the test script I get a noNetwork error:
`Error: could not detect network (event="noNetwork", code=NETWORK_ERROR, version=providers/5.7.2)
`

### Additional Details

I am looking for advice on how others test there scripts easily and quickly without deploying to a public testnet? is the only option to set up `era-test-node`?

**Comment by MexicanAce** [comment, marked as answer]
Hi @GavinStein1 👋 !

There are a few local test strategies that can be used as part of your migration, please see full description in the docs: https://era.zksync.io/docs/tools/testing/#testing-strategies-for-zksync-era

When attempting to use `http://localhost:3050`, that would be against a local node from the [Dockerized Setup](https://era.zksync.io/docs/tools/testing/dockerized-testing.html).

There's also `era_test_node` (as you described) and documentation on using that is available here: https://era.zksync.io/docs/tools/testing/era-test-node.html

At the moment, there's not an "integrated" network like Hardhat's Network Node. So you would need to start up a node in a separate terminal before running your tests.

**Reply by GavinStein1** [reply]
Hi @MexicanAce thanks for the reply!

I think I will configure era_test_node and see how I go then. I look forward to an integrated network being introduced in the future though.

**Reply by caffeinum** [reply]
@GavinStein1 now there's a localnet section on `zksync-cli`:

```shell
bun install -g zksync-cli
bunx zksync-cli localnet up
```
```shell
 Container local-setup-postgres-1  Starting
 Container local-setup-geth-1  Starting
 Container local-setup-geth-1  Started
 Container local-setup-postgres-1  Started
 Container local-setup-zksync-1  Starting
 Container local-setup-zksync-1  Started
```

**Comment by GavinStein1** [comment]
Hi to anyone reading this and is struggling with testing. I have set up the era_test_node using the documentation above and it works quite well. However, I was getting noNetwork errors (sometimes?? go figure) and it is because I was using the `Provider.getDefaultProvider()` function to connect. This function will attempt by default to connect to `http://localhost:3050` and not your test node which is at `http://localhost:8011`. Sometimes the default provider function would connect in certain scopes and I haven't been bothered to find out why. 

What you need to do is use `provider = new Provider('http://localhost:8011')` instead. This will connect to your test node every time. 

@MexicanAce The docs [here](https://era.zksync.io/docs/tools/testing/era-test-node.html#writing-test-scripts) show an example test script for testing on the era_test_node and use the `getDefaultProvider()` function. I think this may cause others to run into similar issues as me. Do you think this should be changed?

**Reply by MexicanAce** [reply]
Hi @GavinStein1 👋

The default provider network is based on values from your `hardhat.config.ts`. Could you double check that the network URL is using port 8011?

If you're still having issues, could you please post your hardhat configuration so we can try to reproduce the error?

**Reply by GavinStein1** [reply]
The network URL is definitely using port 8011. I will share with you my hardhat.config.ts and my testing script that was throwing the noNetwork error.

### Hardhat.config.ts
```
import { HardhatUserConfig } from "hardhat/config";

import "@matterlabs/hardhat-zksync-deploy";
import "@matterlabs/hardhat-zksync-solc";

import "@matterlabs/hardhat-zksync-verify";

// dynamically changes endpoints for local tests
const zkSyncTestnet =
  process.env.NODE_ENV == "test"
    ? {
        url: "http://localhost:8011",
        ethNetwork: "http://localhost:8545",
        zksync: true,
      }
    : {
        url: "https://zksync2-testnet.zksync.dev",
        ethNetwork: "goerli",
        zksync: true,
        // contract verification endpoint
        verifyURL:
          "https://zksync2-testnet-explorer.zksync.dev/contract_verification",
      };

const config: HardhatUserConfig = {
  zksolc: {
    version: "latest",
    settings: {},
  },
  defaultNetwork: "zkSyncTestnet",
  networks: {
    hardhat: {
      zksync: false,
    },
    zkSyncTestnet,
  },
  solidity: {
    version: "0.8.19",
  },
};

export default config;
```

### Test script
```
import { Wallet, Provider, Contract, utils } from 'zksync-web3';
import * as hre from 'hardhat';
import * as ethers from "ethers";
import { Deployer } from '@matterlabs/hardhat-zksync-deploy';
import assert from "assert";
import * as EventContractInfo from "../artifacts-zk/contracts/Event.sol/Event.json";

const RICH_WALLET_PK =
  '0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110';
const EVENT_ABI = EventContractInfo.abi;
const EVENT_BYTECODE = EventContractInfo.bytecode;

// Deploy a contract with constructorArguments
async function deployContract(deployer: Deployer, contractName: string, constructorArguments: any[]): Promise<Contract> {
    const artifact = await deployer.loadArtifact(contractName);
    const overrides = {
        gasLimit: "4294967295"
    }
    return await deployer.deploy(artifact, constructorArguments, overrides);
}


let eventFactoryContract: Contract;
beforeEach(async () => {
    try {
        const provider =  Provider.getDefaultProvider();  // Calling this function here works
        console.log(provider.connection);  // This outputs { url: 'http://localhost:3050' }??
        const account1 = new Wallet(RICH_WALLET_PK, provider);
        const deployer = new Deployer(hre, account1);

        eventFactoryContract = await deployContract(deployer, "EventFactory", []);  // The event factory contract is successfully deployed to the era_test_node
    } catch (err) {
        if (err.code == "NETWORK_ERROR") {
            console.log("No test node found.");  // Not executed when era_test_node is running
        }
    }
    
});

describe("EventFactory", function () {
    describe("Deployment", function () {
        it("Should be deployed", async function () {
            assert(eventFactoryContract.address);
        });
        it("Should have an empty events array", async function () {
            const events = await eventFactoryContract.getEvents();
            assert(events.length == 0);
            });
    });
    describe("New Events", function () {
        it("Should deploy an event", async function () {
            const eventTx = await eventFactoryContract.deployEvent("Event 1");
            const eventRcpt = await eventTx.wait();
            const events = await eventFactoryContract.getEvents();
            const eventAddress = utils.getDeployedContracts(eventRcpt)[0].deployedAddress;
            assert(eventAddress == events[0]);
        });
        it("Should have an event name", async function () {
            const provider = Provider.getDefaultProvider();  // This function here fails to connect to the era_test_node
            const account1 = new Wallet(RICH_WALLET_PK, provider);
            const eventTx = await eventFactoryContract.deployEvent("Event 1");
            const eventRcpt = await eventTx.wait();
            const eventAddress = utils.getDeployedContracts(eventRcpt)[0].deployedAddress;
            const factoryEventAddresses = await eventFactoryContract.getEvents();
            assert(eventAddress == await factoryEventAddresses);

            const eventContract = new ethers.Contract(eventAddress, EVENT_ABI, account1);
            const name = await eventContract.getEventName();  // This function call throws the noNetwork error
            assert(name == 'Event 1');
                        
        });
        it("Should not have tickets", async function () {
            const provider = Provider.getDefaultProvider();
            const account1 = new Wallet(RICH_WALLET_PK, provider);
            const eventTx = await eventFactoryContract.deployEvent("Event 1");
            const eventRcpt = await eventTx.wait();
            const eventAddress = utils.getDeployedContracts(eventRcpt)[0].deployedAddress;
            const eventContract = new ethers.Contract(eventAddress, EVENT_ABI, account1);

            const ticketIDs = await eventContract.getTicketIDs();
            assert(ticketIDs.length == 0);
        });
    });
});
```

Please refer to my comments for more information. Hopefully you are able to reproduce the error with this. Let me know if you want more info.

**Reply by uF4No** [reply]
hey @GavinStein1 

If you use `getDefaultProvider` from the `zksync-web3` package, it'll use a default RPC URL (which is `http://localhost:3050`, the one [used by the docker-setup](https://era.zksync.io/docs/tools/testing/dockerized-testing.html#network-details)) or whatever is defined in the env variable `ZKSYNC_WEB3_API_URL`. Here's the code from the class:

```js
//File: zksync-web3/src/provider.ts

  static getDefaultProvider() {
        // TODO (SMA-1606): Add different urls for different networks.
        return new Provider(process.env.ZKSYNC_WEB3_API_URL || 'http://localhost:3050');
    }
```

However, when using the Deployer class from `hardhat-zksync-deploy` (as you're doing in `const deployer = new Deployer(hre, account1);` the Deployer will initialise a new provider under the hood to interact with the network. This provider will be initialised with the RPC url from the `hardhat.config.ts`. This is not ideal and confusing and we're looking into it!

What can you do? Initialise your provider using the same RPC that you have in the config file. You can hardcode the RPC url or load it from HRE:

```js
// hardcoded RPC URL
const provider = new Provider('http://localhost:8011');
/// RPC from config file
const provider = new Provider(hre.config.networks[hre.config.defaultNetwork].url);
```

Let me know if that helps!

**Reply by GavinStein1** [reply]
Hi @uF4No 

Thanks for the reply! Yes that is super helpful. I ended up hardcoding the RPC url when initialising a provider as you suggested and it works.

**Reply by Zohar97** [reply]
no, hardcode not solve my issue, still getting `noNetwork` error

**Comment by johnwesener** [comment]
To test your zkSync contracts locally without deploying to a public testnet, you can set up a local zkSync node using zkSync's development environment. This environment includes a local testnet called zkSync dev mode. Here's how you can do it:

Install zkSync-cli if you haven't already: 
npm install -g zksync-cli
Initialize a zkSync node:

csharp
zksync init-dev
Start the zkSync dev mode:

zksync dev
Ensure your contracts are configured to use the zkSync dev network. You can do this by setting the network URL to http://localhost:3030 in your deployment scripts or configuration files.

Run your tests as usual, and they will interact with the zkSync dev network.

This setup allows you to quickly and easily test your zkSync contracts locally without the need to deploy to a public testnet. Additionally, you can use tools like Hardhat or Truffle for testing, similar to how you would test contracts on Ethereum's mainnet or testnets.

**Comment by natyfikapsha** [comment]
To test zkSync contracts locally without deploying to a public testnet, you can set up a local zkSync node using zksync-cli. After setting up the local node, configure your contract deployment scripts to connect to the local zkSync node for testing. This allows for rapid development and debugging of zkSync contracts in a local environment.

**Comment by gauvinjohnny** [comment]
To quickly test your contracts without deploying to a public testnet, you can set up a local zkSync test node using the zkSync testnet environment. This allows you to simulate the zkSync environment locally and test your contracts using tools like zkSync CLI or Hardhat plugin. Alternatively, you can utilize zkSync's zkSync Test Helper library to facilitate testing in a local environment. These approaches provide efficient ways to validate contract functionality before deploying to a public testnet or the mainnet.
